---
title: "OctoSeqPipeline"
author: "Mea & Gabby"
date: "8/20/2020"
output: html_document
---
This is the protocol for analyzing Octo Seq data using RStudio (3.6.3), Seurat (3.1.4), Mac (OS Catalina 10.15.3). The first steps outline how to get started with RStudio and Seurat and how to set up sshfs for Mac in order to access raw data from the Talapas server. If you have already set up R, RStudio, Seurat, and sshfs, then you can begin at section "Mount to Talapas" and proceed with setting your working directory and reading in the raw data.

## Download Seurat packages (source: https://satijalab.org/seurat/install.html) ~15m
# Enter commands in R or RStudio
```{r}
install.packages('BiocManager')
BiocManager::install('multtest') 
# Do you want to install from sources the packages which need compilation? (y/n/cancel); update a/s/n? a
install.packages('Seurat')
# package which is only available in source form, and may need compilation? (y/n) Yes
install.packages("sn")
library(Seurat)
# Would you like to install Miniconda? (y/n) Y
```
## Download FUSE for macOS to be able to mount to Talapas using sshfs
Go to this website: https://osxfuse.github.io
Follow instructions and allow Benjamin Fisher access through System Preferences
Install Brew packagemanager (source: https://brew.sh): in terminal command, enter: /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)" 
In terminal command, enter: brew install sshfs 

# Mount to talapas
In terminal command, enter: sshfs gcoffing@talapas-ln1.uoregon.edu:/projects/niell/shared/Octopus /Users/gcoffing/Octopus 
[In terminal command, enter: sshfs jsongco@talapas-ln1.uoregon.edu:/projects/niell/shared/Octopus /Users/gcoffing/Octopus]
Prior to mounting, user will need to create a directory to mount to; in my case, I created the directory "Octopus"

## Set working directory (run this line of code from the console rather than in a chunk)
```{r}
setwd("/Users/gcoffing/Documents/Documents/octo/Seurat")
setwd("/Users/josongco/Desktop/ABD/FinalSeqData")
```

## Load libraries for Seurat
```{r}
library(Seurat)
library(Matrix)
library(ggplot2)
library(sctransform)
library(stringr)
```

## Set pathway & parameters
```{r}
# change depending on user
user = "mea"
if (user == "mea"){
  directoryMaster = "~/Google Drive/miller niell octo seq/project data/scRNAseq/Seurat/directoryMaster_072821.csv"
  full_namekey = "~/Google Drive/miller niell octo seq/project data/scRNAseq/Seurat/gene lists/geneIDs_miniNCBI_fulldf_namekey_Sep18_2.csv"
} else if (user == "gabby"){
  directoryMaster = "directoryMaster_070720.csv"
  full_namekey = "fulldf_namekey.csv"
}
```

## Read in files
```{r}
# read in csv with the directories
datafiles <- read.csv(directoryMaster, stringsAsFactors = FALSE) 
data_dir <- as.vector(datafiles$Directory)

# read in csv namekey
namekey <- read.csv(full_namekey, stringsAsFactors = FALSE)
```

## Hashmap for replacing names: Original or Hisat GTF?
```{r}
#making a hashmap of OctoGenes/Ocbims to fullnames from the seurat namekey 
gtf = "hisat"
if (gtf == "original") {
    unclass(by(namekey, namekey$Ocbimv, function(x) {
     tmp <- x$fullname
      setNames(tmp, x$Ocbimv[1])
     tmp
     })) -> namekeyDict
} else if (gtf == "hisat"){
   unclass(by(namekey, namekey$OctoGene, function(x) {
     tmp <- x$fullname
     setNames(tmp, x$OctoGene[1])
      tmp
      })) -> namekeyDict
}
```

## Check that hashmap worked properly
```{r}
if (gtf == "original") {
#Check that the hashmap is working properly (original gtf)
print(namekeyDict[["Ocbimv22027788m"]]) #should print  "NA-OG10040-OB27788"
print(namekeyDict[["Ocbimv22030350m"]]) #should print "Endonuclease-Exonuclease-phosphatase-family-phosphodiesterase-12-OG10002-OB30350" 
} else if (gtf == "hisat") {
#Check that the hashmap is working properly (hisat)
print(namekeyDict[["OCTOGene.1"]]) #should print "NA-OG1"
print(namekeyDict[["OCTOGene.10032"]]) #should print "anaphase-promoting-complex-subunit-7-OG10032-OB20676-OB29021"
}
```

## An overview of data processing is as follows:
1. Read in raw data
2. Replace names with gene IDS
3. Create a Seurat object for downstream analyses and data manipulation

# Using a for loop, load in Octo Seq Data and take data through pre-processing steps
```{r}
#make empty list for all the seurat objects
all_list = list() 

ct = 1 #what is ct?
for(f in 1:2){ #Old or New gtf, NextSeq + HiSeq from directory master, ipa 
  all <- Read10X(data.dir = data_dir[f+22]) # f+16 for new gtf HiSeq, f+10 for old gtf HiSeq, f+18 for ipa gtf 1 & 2, f+22 for new hisat gtf + transdecoder
    ngenes <- length(all@Dimnames[[1]])
    if (gtf == "original"){ #replacing Ocbim names
      for (g in 1:ngenes){
        gene <- all@Dimnames[[1]][g]
        gene <- substr(gene, 6, str_length(gene)-2)
        ind<- grep(gene, namekey[[3]])
        if (length(ind)>0) {
          id <- namekey[[ind[1], 2]]
          if (str_length(id)>0){
            id <- str_remove_all(id, "\\(") # parentheses mess up gene names as dimensions
            id <- str_remove_all(id, "\\)") 
            id <- substr(id, 1, 60) # keep it short
            all@Dimnames[[1]][g] <- temp
        }
      }
        } 
      } else if (gtf == "hisat"){ #replacing Octogene names
      for (g in 1:ngenes){
        gene <- all@Dimnames[[1]][g]
        temp <- namekeyDict[[gene]]
        if (length(temp) == length(all@Dimnames[[1]][g])){ #This checks that temp is not zero
         all@Dimnames[[1]][g] <- temp
         }
       }
      }
  all <- CreateSeuratObject(counts = all, project = paste("Sample", f, sep = ""), min.cells = 3, min.features = 750)
  all_list[[ct]] = all
  ct = ct + 1
}


# Check that all names were replaced correctly # does not work for original gtf
grep("VGlut",all_list[[1]]@assays[["RNA"]]@data@Dimnames[[1]], ignore.case = TRUE) # prints [1] 2646 5437
# double check that this is the right gene that you were looking for
all_list[[1]]@assays[["RNA"]]@data@Dimnames[[1]][5437] # prints "VGlut-Major-Facilitator-Superfamily-solute-carrier-family-17-sodium-dependent-inorganic-phosphate-cotransporter-member-6-OG8038-OB24718"

saveRDS(all_list, "/Users/josongco/Desktop/ABD/FinalSeqData/ipa_All_list_noname062121.rds") #change depending on dataset
  
# To access individual objects: all_list[[1]]
#example: all_list[[1]] will print the object information on Seurat object 1
#
#
#

ngenes <- length(all@Dimnames[[1]])
    if (gtf == "original"){ #replacing Ocbim names
      for (g in 1:ngenes){
        gene <- all@Dimnames[[1]][g]
        gene <- substr(gene, 6, str_length(gene)-2)
        ind<- grep(gene, namekey[[3]])
        if (length(ind)>0) {
          id <- namekey[[ind[1], 2]]
          if (str_length(id)>0){
            id <- str_remove_all(id, "\\(") # parentheses mess up gene names as dimensions
            id <- str_remove_all(id, "\\)") 
            id <- substr(id, 1, 60) # keep it short
            all@Dimnames[[1]][g] <- temp
        }
      }
        } 
      } else if (gtf == "hisat"){ #replacing Octogene names
      for (g in 1:ngenes){
        gene <- all@Dimnames[[1]][g]
        temp <- namekeyDict[[gene]]
        if (length(temp) == length(all@Dimnames[[1]][g])){ #This checks that temp is not zero
         all@Dimnames[[1]][g] <- temp
         }
       }
      }

```

