---
title: "Straight merge of v3 chemistry (std norm)"
author: "jos"
date: "05/11/2020"
output:
  html_document:
    fig_height: 12
    fig_width: 15
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Objective: use a for loop to read in and process v3 data sets from the original GTF and conduct a straight merge with standard normalization.

## Read in data
Using the OctoSeqPipeline Code, I used a for loop to read in the v3 datasets from the original GTF. I set parameters to include a min.cell = 3 and min.features = 200. 
```{r , echo = FALSE}
# to knit document, use the following code: rmarkdown::render("/Users/josongco/Documents/GitHub/Octopolis/OSall_30pcs30dimsres1_normSTR8merge_v3.Rmd")
library(Seurat)
library(Matrix)
library(ggplot2)
library(sctransform)
library(stringr)
```

## Read in normalized datasets
I performed normalization on the individual datasets before conducting a straight merge. I used the "merge" function to create a new Seurat object "all" for downstream analyses but preserved the original identities of the individual datasets via the merge argument "add.cell.ids".
```{r}
#for (i in 2:length(all_list)){
 # all_list[[i]][["percent.mt"]] <- PercentageFeatureSet(all_list[[i]], pattern = "^mitochondria-") #preferred for for loop over original command
  #all_list[[i]] <- NormalizeData(all_list[[i]])
#}

#all <- merge(x = all_list[[2]], y = all_list[[3]], add.cell.ids = c("Run2", "Run3"), project = "StdN_str8_v3")
#saveRDS(all, "/Users/josongco/Desktop/SeqData/allnormV3.rds")

all <- readRDS("/Users/josongco/Desktop/SeqData/allnormV3.rds")
```

## Conduct a straight merge and plot features of the object. 
I used visualization techniques to plot nFeature, nCount, and percent.mt for each of the Runs in the merged object. (Note: nFeature = # genes, nCount = UMI counts, and percent.mt is important because higher levels of percent.mt are indicative of cell stress.) Typically, I would subset the data to only include nFeatures >200, <2500, and percent.mt <5. However, even though I used visualization techniques to plot these features, I did not end up subsetting the data.
```{r }
VlnPlot(all, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(all, feature1 = "nCount_RNA", feature2 = "percent.mt")
# visualize the relationship between UMI counts and mitochondrial genes (percent)
plot2 <- FeatureScatter(all, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
# visualize the relationship between genes (ID as >1 UMI) and mitochondrial genes
plot1 + plot2
```

## Determine variable features
I then found variable features and plotted the top10 variable features.
```{r }
all <- FindVariableFeatures(all, selection.method = "vst", nfeatures = 2000)
top10 <- head(VariableFeatures(all), 10) # ID 10 most highly variable genes
plot1 <- VariableFeaturePlot(all)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
print(plot1)
print(plot2)
```

## Run PCA and generate UMAP
Next, I scaled the data. I regressed out only percent.mt. I then performed PCA and all downstream analyses on 30 pcs, 30 dims, and a resolution of 1. I then plotted the UMAP based on this original analysis and then again with the Runs mapped on top of the clusters. We're able to see that a straight merge does not integrate the data as well as CCA does. 
```{r }
all <- ScaleData(all, vars.to.regress = "percent.mt") 
# unable to regress out nCount_RNA, which causes the ScaleData 
# to not properly store information which is necessary for RunPCA function
all <- RunPCA(all, npcs = 30, verbose = FALSE) 
all <- FindNeighbors(all, reduction = "pca", dims = 1:30)
all <- FindClusters(all, resolution = 1)
all <- RunUMAP(all, reduction = "pca", dims = 1:30)
DimPlot(all, reduction = "umap", label = TRUE)
DimPlot(all, reduction = "umap", group.by = "orig.ident")
```

## Examine cell-cluster information
I also extracted information about cell#-cluster relationships. In short, we're able to see how many cells are in each cluster, how many cells are in each replicate, what proportion of cells are in each cluster, and how cluster membership varies by replicate. I also extracted this information after going through the hierarchical clustering steps and reassigning numbers based on the dendogram to see if there are any differences.
```{r }
table(Idents(all))
table(all$orig.ident)
prop.table(table(Idents(all)))
table(Idents(all), all$orig.ident)
```
## Build a hierarchical tree
I then followed the pipeline for building a hierarchical tree and created a new Seurat object "all.tree". I generated a UMAP based on the tree and extracted the cell#-cluster relationship information from this UMAP as well. 
```{r }
all.tree <- BuildClusterTree(all, reorder = TRUE, reorder.numeric = TRUE, slot = "scale.data", verbose = TRUE, dims = 1:30)
PlotClusterTree(all.tree, label = TRUE)
```

```{r }
all.tree <- RunUMAP(all.tree, dims = 1:30)
DimPlot(all.tree, reduction = "umap", label = TRUE)
table(Idents(all.tree)) 
table(all.tree$orig.ident) 
prop.table(table(Idents(all.tree))) 
table(Idents(all.tree), all.tree$orig.ident) 
```

## Plot features
I then selected a list of genes (in situ) to use as verification of how well the different parameters (i.e standard norm, straight merge vs CCA merge) are working. 
````{r, echo = FALSE}
yfg <- read.csv("~/GDrive/miller niell octo seq/project data/scRNAseq/Seurat/gene lists/Genes for in situ.csv",stringsAsFactors=FALSE)

all.genes <- rownames(all) # same as rownames(all.tree)
#all.genes <- rownames(all@assays$RNA) 

genelist <- vector()
nomatch <- list()
for (i in 1:12){
  gene <- yfg[[i,2]]
  #gene<-substr(gene,7,str_length(gene)-1)
  loc <- grep(gene,all.genes)
  if (length(loc)>0) {
    genelist <- c(genelist,loc)
  } else {
    nomatch <- c(nomatch,yfg[[i,2]])
  }
}

DotPlot(all.tree,features=rev(all.genes[genelist[1:12]])) + RotatedAxis()

FeaturePlot(all.tree,features = all.genes[genelist[1:2]], ncol = 1) + NoLegend() + NoAxes()
FeaturePlot(all.tree,features = all.genes[genelist[3:4]], ncol = 1) + NoLegend() + NoAxes()
FeaturePlot(all.tree,features = all.genes[genelist[5:6]], ncol = 1) + NoLegend() + NoAxes()
FeaturePlot(all.tree,features = all.genes[genelist[7:8]], ncol = 1) + NoLegend() + NoAxes()
FeaturePlot(all.tree,features = all.genes[genelist[9:10]], ncol = 1) + NoLegend() + NoAxes()
FeaturePlot(all.tree,features = all.genes[genelist[11:12]], ncol = 1) + NoLegend() + NoAxes()

VlnPlot(all.tree,features = all.genes[genelist[1:4]], ncol = 1) + NoLegend() + NoAxes()
VlnPlot(all.tree,features = all.genes[genelist[5:8]], ncol = 1) + NoLegend() + NoAxes()
VlnPlot(all.tree,features = all.genes[genelist[9:12]], ncol = 1) + NoLegend() + NoAxes()

# Can also use line of code below to find candidate gene 
#FeaturePlot(all.tree,features = all.genes[grep("VGlut",all.genes)])