---
output:
  pdf_document: default
  html_document: default
---
# VPNconnect: vpn.uoregon.edu, login ID

## Mount to Talapas to access raw data (via Terminal Window)
#sshfs dniell@talapas-ln1.uoregon.edu:/projects/niell/dniell /Users/deniseniell/talapas_dniell
#cd /Users/deniseniell/talapas_dniell/cellranger


# Open Rmd file (the following commands will be in RStudio)
## Set your working directory path
```{r}
#setwd("/Users/deniseniell/Desktop/Seurat/run2")
rm(list = ls())
```

# Load libraries 
```{r}
library(dplyr)
#Note: you might get an error when loading dplyr but if you just run the command again, then it will load and everything should work well
library(Seurat)
library(Matrix)
library(ggplot2)
library(sctransform)
library(stringr)
```

# Load the Octo Seq raw data ~2m
Keep track of which data set you are working with. The code is written so that you can run through the pipeline without having to worry about changing variables (i.e. data sets are imported to the variable "all" so that all the following commands will process the data without you needing to change the variable each time)
```{r}
all <- Read10X(data.dir = "D:/data/octo seq/Cellranger/OctoSeq2.1/raw_feature_bc_matrix")
```

# replace gene names in raw data with IDs from ccsv file
# takes a few minutes

```{r}
ref <- read.csv("D:/data/octo seq/refMaster_040420.csv",stringsAsFactors=FALSE)

ngenes <- length(all@Dimnames[[1]])
for (g in 1:ngenes){
  gene<-all@Dimnames[[1]][g]
  gene<-substr(gene,6,str_length(gene)-2)
  ind<-grep(gene,ref[[1]])
  if (length(ind)>0) {
    id <- ref[[ind[1],2]]
    if (str_length(id)>0) {
      id <- str_remove_all(id,"\\(") # parentheses mess up gene names as dimensions
      id <- str_remove_all(id,"\\)")
      id <- substr(id,1,60) # keep it short
      all@Dimnames[[1]][g]<- paste(id,gene,sep='-')
    }
  }
}
```
# option to read in RDS file here
```{r}
#all <- readRDS(file = "/Users/deniseniell/Desktop/Seurat/run2/OSmarkersTree.rds")
```

# Initialize the Seurat object with the raw (non-normalized data).
```{r}
#Change the project name here so that you can keep track of your data

all <- CreateSeuratObject(counts = all, project = "OctoSeq2_names", min.cells = 3, min.features = 200)
all
```

# Begin Quality Control Steps
```{r}
mito.genes <- grep(pattern = "^mt-", x = rownames(x = all), value = TRUE)
percent.mito <- Matrix::colSums(all) / Matrix::colSums(all)
all[["percent.mt"]] <- PercentageFeatureSet(all, pattern = "^MT-")
all <- subset(all, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
plot1 <- FeatureScatter(all, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(all, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
CombinePlots(plots = list(plot1, plot2))
```

## Number of unique genes and total moleculares are automatically calculated during CreateSeuratObject, but you can view these metrics because they are stored in the object meta data
```{r}
head(all@meta.data, 5)
```
## You can also visualize these QC metrics and use this to decide how to filter cells
```{r}
VlnPlot(all, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3)
```
# Begin normalizing the data
```{r}
all <- NormalizeData(all, normalization.method = "LogNormalize", scale.factor = 10000)
```
## You can also use the following command and you should receive the same output
```{r}
#all <- NormalizeData(all)
```

# Identification of highly variable features (feature selection), and plot these features
```{r}
all <- FindVariableFeatures(all, selection.method = "vst", nfeatures = 2000)
```
# Identify the 10 most highly variable genes
```{r}
top10 <- head(VariableFeatures(all),10)
```
# Plot variable features with and without labels
```{r}
plot3 <- VariableFeaturePlot(all)
plot4 <- LabelPoints(plot = plot3, points = top10, repel = TRUE)
#CombinePlots(plots = list(plot3, plot4))
```
# Apply a linear transformation (scaling), which is a standard pre-processing step prior to dimensional reduction techniques like PCA ~2m
```{r}
all.genes <- rownames(all)
all <- ScaleData(all, features = all.genes)
```
# Perform linear dimensional reduction (PCA) ~15-20m
I normally will create a new variable/Seurat object at this point and rename it as "all.norm" so that I know that I have all of the preprocesed data prior to running PCA, and I can manipulate the PCs etc moving forward but still be able to go back to the processed data easily
```{r}
all.norm <- RunPCA(all, features = VariableFeatures(object = all), npcs = 200)
```
# save data at this point just in case RStudio crashes!
```{r}
#saveRDS(all.norm, file = #"/Users/deniseniell/Desktop/Seurat/run2/OSr2norm200.rds")
```

# Visualize PCAs in a few different ways
You can skip the next few lines of code (visualization of PCs, elbow and jackstraw) if you know how many pcs you want to move forward with.
```{r}
DimHeatmap(all.norm, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(all.norm, dims = 1:10, cells = 500, balanced = TRUE)
DimHeatmap(all.norm, dims = c(90:100, 197:200), cells = 500, balanced = TRUE)
VizDimLoadings(all.norm, dims = 1:2, reduction = "pca")
DimPlot(all.norm, reduction = "pca")
```
# Seurat clusters cells based on their PCA scores, with each PC essentially representing a "metafeature" that combines information across a correlated feature set. Top principle components represent robust compression of the dataset. To determine how many components to include, one can utilize a resampling test inspired by the JackStraw procedure.

## Use the ElbowPlot function before JackStraw to visualize and save on computing power since the JackStraw method can take a long time for processing big datasets, others rely on the ElbowPlot function to reduce computation time and still produce an approximation.
```{r}
ElbowPlot(all.norm, ndims = 200)
```

## This sampling strategy randomly permutes a subset of the data (1% is the default) ad erun PCA, constructing a "null distribution" of feature scores, and repeat this procedure. This allows an identification of "significant" PCs as those who have a strong enrichment of low p-value features.

### The JackStrawPlot function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (indicated by the dashed line). "Significant" PCs will show a strong enrichment of features with low p-values (indicated by the solid curve above the dashed line). ~30m

##Note: even if you run 200pcs, the max # of dims you can use for ScoreJackStraw and JackStrawPlot is 20.
```{r}
#all.norm <- JackStraw(all.norm, num.replicate = 100)
#all.norm <- ScoreJackStraw(all.norm, dims = 1:20)
#JackStrawPlot(all.norm, dims = 1:20)
```
# Cluster the cells

## Seurat v3 applies a graph-based clustering approach. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected "quasi-cliques" or "communities"

### (like PhenoGraph), a KNN graph is constructed based on the elucidean distance in PCA space. Then, the FindNeighbors function refines edge weights between any two cells based on the shared overlap in their local neighborhoods (this is referred to as Jaccard similarity). This function takes the previously defined input regarding number of PCs (now, dims)

## To cluster cells, Seurat applies a modularity optimization technique, such as Louvain algorithm (which is the default) or SLM, to iteratively group cells together, with the goal of optimizing the standard modularity function. 

### FindClusters implements the procedure mentioned above and contains a resolution parameter that sets "granularity" of downstream clustering, with increased values leading to a greater number of clusters. Typically, setting the parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Clusters could be found using the Idents function.

##Note: even if you only run 20 dims with the JackStraw function above, looks like you can still proceed with >20 dims in the following commands ~5m
```{r}
all.norm <- FindNeighbors(all.norm, dims = 1:50) # changed this from 200, to be consistent with FindClusters next
all.norm <- FindClusters(all.norm, reduction.type = "pca", dims = 1:50, resolution = 1)
##Notes: 200dims, res 1 yields 18 singletons and 15 final clusters; 200dims, res 0.5 yields 18 singletons and 12 final clusters; 200dims, res 1.5 yields 29 singletons and 68 final clusters; 200dims, res 0.2 yields 18 singletons and 9 final clusters; 150dims, res 1 yields 18 singletons and 15 final clusters; 100dims, res 1 yields 18 singletons and 15 final clusters; 50dims, res 1 yields 18 singletons and 15 final clusters; 50dims, res 0.5 yields 18 singletons and 12 clusters. --> let's stick with 50dims at res 1 for now and revisit FindNeighbors parameters to address singleton issue
##Notes: Neighbors = 200 dims, Clusters = 50dims, at res 0.05 there are 18 singletons and 9 final clusters; Neighbors = 200, Clusters = 50dims, at res 0.01 there are 18 singletons and 6 final clusters

## Additional notes: for OctoSeq2, 50dims res 1 yields 1 singleton and 21 final clusters.
```
# Look at cluster IDS of the first 5 cells
```{r}
head(Idents(all.norm), 5)
```
# Run non-linear dimensional reduction (UMAP/tSNE)

## Non-linear dimensional reduction techniques, such as tSNE and UMAP, allows one to visualize and explore the datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in a low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. Seurat suggests using the same PCs as what was provided as input to the clustering analysis.
```{r}
all.norm <- RunUMAP(all.norm, dims = 1:50) # was 1:20 dims
DimPlot(all.norm, reduction = "umap", label = TRUE)
```
# Run non-linear dimensional reduction with tSNE (UMAP preferred, so this section is edited/commented out)
#OS1.15 <-RunTSNE(object = OS1.norm, dims = 1:15, do.fast = TRUE)
TSNEPlot(object = OS1.15, do.label = TRUE)

```{r}
#saveRDS(all.norm, file = #"/Users/deniseniell/Desktop/Seurat/run2/OSr2PCs200res50.rds")
```

# Build a phylogenetic tree, and rename/reorder cluster names according to their position on the tree
# See help for details on tree building strategy
# This gives closely related clusters similar cluster IDs, which is occasionally useful for visualization later on
# Assigned cluster will be placed in the 'tree.ident' field of nbt@data.info, and also stored in nbt@ident

```{r}
all.normTREE <- BuildClusterTree(all.norm, reorder = TRUE, reorder.numeric = TRUE, slot = "scale.data", verbose = TRUE, dims=1:50)
PlotClusterTree(all.normTREE)
```

# Run non-linear dimensional reduction (UMAP/tSNE)

## Non-linear dimensional reduction techniques, such as tSNE and UMAP, allows one to visualize and explore the datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in a low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. Seurat suggests using the same PCs as what was provided as input to the clustering analysis.

```{r}
# don't need to recalculate, just replot
#all.normTREE <- RunUMAP(all.normTREE, dims = 1:50)
DimPlot(all.normTREE, reduction = "umap", label = TRUE)
```



# Finding differentially expressed features (for 6 clusters, ~15m)
#FindMarkers will find markers between two different identity groups - you have to specify both identity groups. This is useful for comparing the differences between two specific groups.

#FindAllMarkers will find markers differentially expressed in each identity group by comparing it to all of the others - you don't have to manually define anything. Note that markers may bleed over between closely-related groups - they are not forced to be specific to only one group. This is what most people use (and likely what you want).

```{r}
cluster.markers <- FindAllMarkers(all.normTREE, min.pct = 0.5, logfc.threshold = 0.5)
#write.csv(cluster.markers, file = "/Users/deniseniell/Desktop/Seurat/run2/clustermarkers_OSr2PCs200res50.csv")
```

```{r}
#saveRDS(all.normTREE, file = #"/Users/deniseniell/Desktop/Seurat/run2/OSmarkersTree.rds")
```

# Visualizing top 10 markers
```{r}
top10 <- cluster.markers %>% group_by(cluster) %>% top_n(n = 5,wt = avg_logFC )
#top10 <- cluster.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
#DoHeatmap(all.normTREE, features = top10$gene) + NoLegend() + theme(axis.text.y = element_text(size = 5))
DotPlot(all.normTREE,features=rev(unique(top10[1:50,]$gene))) + RotatedAxis()+ theme(axis.text.x = element_text(size = 7))
DotPlot(all.normTREE,features=rev(unique(top10[51:100,]$gene))) + RotatedAxis()+ theme(axis.text.x = element_text(size = 7))
DotPlot(all.normTREE,features=rev(unique(top10[101:159,]$gene))) + RotatedAxis()+ theme(axis.text.x = element_text(size = 7))


```


## get markers for each cluster separated out
```{r}
cluster_markers <- list()

nclust = nlevels(Idents(all.normTREE))
for(i in 1:nclust) {
cluster_markers[[i]] <- cluster.markers[which(cluster.markers$cluster == i),]
   #cluster_markers[[i]] <- FindMarkers(
    # all.normTREE, ident.1 = i, min.pct=0.5, logfc.threshold = 0.5)
}
```

## heatmap for individual clusters
```{r}

for (i in 1:nclust){
  these <- cluster_markers[[i]]
  these <-these[which(these$avg_logFC>0.5),]
  these <- these[order(these$avg_logFC, decreasing = TRUE),]
  
   print(DoHeatmap(all.normTREE, features = head(these$gene,40)) + NoLegend() + theme(axis.text.y = element_text(size = 5)))

   print(head(these,40))
}
```
### get markers for pairwise discrimination
```{r}
pair.markers <- FindMarkers(all.normTREE, ident.1 = 24, ident.2=25,min.pct = 0.5, logfc.threshold = 0.5)

   DoHeatmap(all.normTREE, features = rownames(pair.markers)) + NoLegend() + theme(axis.text.y = element_text(size = 8))
```
# get markers of groups
```{r}

pair.markers <- FindMarkers(all.normTREE, ident.1 = c(15,16,17, 18, 19),min.pct = 0.5, logfc.threshold = 0.5)

   DoHeatmap(all.normTREE, features = rownames(pair.markers)) + NoLegend() + theme(axis.text.y = element_text(size = 8))
```


#tree maps--------------------------------------------------------------------------

# Grab possible nodes using FindMarkers from seurat object, assuming you've already run BuildClusterTree()
```{r}
nodes <- unique(all.normTREE@tools$BuildClusterTree$edge[,1])

tree_markers <- list()
for(i in 1:length(nodes)) {
#for(i in 1:2) {
   tree_markers[[i]] <- FindMarkers(
     all.normTREE, ident.1 = "clustertree", ident.2 = nodes[i], min.pct=0.5)
}

```

# selecting markers that are greater than log2, sorting them and taking 10 most neg (left?) and 10 most pos (right?)
```{r}
goodmarkers <- list()
leftMarkers <- list()
rightMarkers <- list()
for(i in 1:length(nodes)){
  these <- tree_markers[[i]]
  these <-these[which(abs(these$avg_logFC)>0.5),]
  these <- these[order(these$avg_logFC, decreasing = TRUE),]
  these$node_id <- nodes[[i]]
  these$gene_id<-rownames(these)
  leftMarkers[[i]] <- head(these,3)
  rightMarkers[[i]] <- tail(these,3)
  goodmarkers[[i]] <- these

}

```
#save out goodmarkers[[i]] and nodes to csv file


```{r}

#rbind.fill(goodmarkers)
#df_goodtable <- ldply(goodmarkers, data.frame)
#capture.output(summary(df_goodtable), file = "/Users/deniseniell/Desktop/Seurat/run2/R2goodtable")
```

```{r}
#write.csv(df_goodtable, file = "/Users/deniseniell/Desktop/Seurat/run2/R2goodmark_list.csv")


```


# concatenate left and right markers for all nodes
```{r}
allmarkers <- rownames(goodmarkers[[1]])
leftRightMarkers <- c(rownames(leftMarkers[[1]]),rownames(rightMarkers[[1]]))
for(i in 2:length(nodes)){
  allmarkers <- c(allmarkers,rownames(goodmarkers[[i]]))
  leftRightMarkers <- c(leftRightMarkers,rownames(leftMarkers[[i]]),rownames(rightMarkers[[i]]))
  
}
```

#heatmap for each node individually
```{r}
for(i in 1:length(tree_markers)){
  i
  these <- tree_markers[[i]]
  these <-these[which(abs(these$avg_logFC)>0.5),]
  these <- these[order(these$avg_logFC, decreasing = TRUE),]
  these <- rbind(head(these,10),tail(these,10))
print(these)
   print(DoHeatmap(all.normTREE, features = rownames(these))+ NoLegend() + theme(axis.text.y = element_text(size = 5)))
  
}

 
```

#dotplot for one node individually
```{r}
i<-14
  i
  these <- tree_markers[[i]]
  these <-these[which(abs(these$avg_logFC)>0),]
  these <- these[order(these$avg_logFC, decreasing = TRUE),]
  these <- rbind(head(these,10),tail(these,10))

DotPlot(all.normTREE,features=rownames(these))+ RotatedAxis()+ theme(axis.text.x = element_text(size = 6))
  
```


#Heatmap of all concatenated node data
```{r}
DoHeatmap(all.normTREE, features = leftRightMarkers) + NoLegend() + theme(axis.text.y = element_text(size = 4))

```

## map developmental genes
```{r}
genelist <- vector()
nomatch <- list()
for (i in 1:129){
  gene <- ref[[i,1]]
  gene<-substr(gene,7,str_length(gene)-1)
  loc <- grep(gene,all.genes)
  if (length(loc)>0) {
    genelist <- c(genelist,loc)
  } else {
    nomatch <- c(nomatch,ref[[i,2]])
  }
}
DoHeatmap(all.normTREE, features = all.genes[genelist],disp.min = -1.5, disp.max = 1.5)  + theme(axis.text.y = element_text(size = 4))

```

# map neural genes
```{r}

genelist <- vector();
nomatch <- list();
for (i in 130:207){
  gene <- ref[[i,1]]
  gene<-substr(gene,7,str_length(gene)-1)
  loc <- grep(gene,all.genes)
  if (length(loc)>0) {
    genelist <- c(genelist,loc)
  } else {
    nomatch <- c(nomatch,ref[[i,2]])
  }
}
DoHeatmap(all.normTREE, features = all.genes[genelist], disp.min = -1.5, disp.max = 1.5)  + theme(axis.text.y = element_text(size = 5))

```
#map cadherins
```{r}
DoHeatmap(all.normTREE, features = all.genes[grep("Cadherin-O",all.genes)], disp.min = -1.5, disp.max = 1.5)  + theme(axis.text.y = element_text(size = 8))
```

#heatmap of yfg (your favorite gene)
```{r}
yfg <- read.csv("D:/data/octo seq/Genes for in situ.csv",stringsAsFactors=FALSE)

genelist <- vector()
nomatch <- list()
for (i in 1:22){
  gene <- yfg[[i,2]]
  #gene<-substr(gene,7,str_length(gene)-1)
  loc <- grep(gene,all.genes)
  if (length(loc)>0) {
    genelist <- c(genelist,loc)
  } else {
    nomatch <- c(nomatch,yfg[[i,2]])
  }
}
DoHeatmap(all.normTREE, features = all.genes[genelist],disp.min = -1.5, disp.max = 1.5)  + theme(axis.text.y = element_text(size = 8))




```

```{r}
DotPlot(all.normTREE,features=rev(all.genes[genelist])) + RotatedAxis()
```
```{r}

#heatmap of yfg (your favorite gene)

yfg <- read.csv("D:/data/octo seq/GeneIDs - More Neuro.csv",stringsAsFactors=FALSE)
yfg.unique = unique(yfg[,2])
genelist <- vector()
nomatch <- list()
for (i in 1:length(yfg.unique)){
  gene <- yfg.unique[i]
#gene<-substr(gene,1,str_length(gene)-1)
  loc <- grep(gene,all.genes)
  if (length(loc)>0) {
    genelist <- c(genelist,loc)
  } else {
    nomatch <- c(nomatch,yfg[[i,2]])
  }
}
DotPlot(all.normTREE,features=rev(all.genes[genelist[1:100]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=rev(all.genes[genelist[101:200]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=rev(all.genes[genelist[201:270]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))



```
```{r}

#heatmap of gpcrs

yfg <- read.csv("D:/data/octo seq/GeneIDs - All GPCRs.csv",stringsAsFactors=FALSE)

genelist <- vector()
nomatch <- list()
for (i in 1:327){
  gene <- yfg[[i,2]]
#gene<-substr(gene,1,str_length(gene)-1)
  loc <- grep(gene,all.genes)
  if (length(loc)>0) {
    genelist <- c(genelist,loc)
  } else {
    nomatch <- c(nomatch,yfg[[i,2]])
  }
}
DotPlot(all.normTREE,features=rev(all.genes[genelist[1:50]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=rev(all.genes[genelist[51:100]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=rev(all.genes[genelist[101:150]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=rev(all.genes[genelist[151:200]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=rev(all.genes[genelist[201:243]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

```


```{r}
zinc <- grep("c2h2",all.genes,ignore.case=TRUE)
DotPlot(all.normTREE,features=rev(all.genes[zinc[1:500]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 4))

DotPlot(all.normTREE,features=rev(all.genes[zinc[501:1000]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 4))

DotPlot(all.normTREE,features=rev(all.genes[zinc[1001:1500]])) + RotatedAxis()+ theme(axis.text.x = element_text(size = 4))

```
```{r}
#krab c2h2 zinc fingers are noted in albertin et al. probably not getting them all
DotPlot(all.normTREE,features=all.genes[grep("krab",all.genes,ignore.case=TRUE)]) + RotatedAxis()+ theme(axis.text.x = element_text(size = 4))
```
```{r}
DotPlot(all.normTREE,features=all.genes[grep("hox",all.genes,ignore.case=TRUE)]) + RotatedAxis()+ theme(axis.text.x = element_text(size = 4))
```

```{r}
cadh <-grep("cadherin",all.genes,ignore.case=TRUE)


DotPlot(all.normTREE,features=all.genes[cadh[1:50]]) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=all.genes[cadh[51:100]]) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=all.genes[cadh[101:150]]) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

DotPlot(all.normTREE,features=all.genes[cadh[151:193]]) + RotatedAxis()+ theme(axis.text.x = element_text(size = 6))

```

```{r}
FeaturePlot(all.normTREE,features = all.genes[grep("protocadherin",all.genes,ignore.case=TRUE)])
print(nomatch)

FeaturePlot(all.normTREE,features = all.genes[genelist[13:22]]) + NoLegend() + NoAxes()

FeaturePlot(all.normTREE,features = all.genes[grep("Synaptotagmin",all.genes)],max.cutoff = 10)
```

```{r}
FeaturePlot(all.normTREE,features = all.genes[grep("021175",all.genes)]) + 
scale_color_gradientn( colours = c('lightgrey', 'blue'),  limits = c(0, 8))

FeaturePlot(all.normTREE,features = all.genes[grep("VACHT",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("VGlut",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("TH-O",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("TyrBH",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("FMRF amide",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("FMRF related",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("000748",all.genes)])

FeaturePlot(all.normTREE,features = all.genes[grep("25965",all.genes)])



```
```{r}
FeaturePlot(all.normTREE,features = all.genes[grep("glutsyn",all.genes,ignore.case=TRUE)])
```

